tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -101,39 +101,100 @@ bool IsF8Type(const HloInstruction *instr) {
   }
 }
 
-bool IsF8TypeRecursiveImpl(const HloInstruction *instr,
-                           absl::flat_hash_set<int> &visited_instrs) {
+// Recursively collects unary, pad, divide or multiply operands of instr until
+// an instruction with FP8 element type is reached. Returns std::nullopt when no
+// FP8 instruction is reached.
+std::optional<std::vector<HloInstruction *>> FindF8SubgraphRecursive(
+    const HloInstruction *instr, absl::flat_hash_set<int> &visited_instrs,
+    std::vector<HloInstruction *> subgraph) {
   // Avoid visiting the same instruction more than once.
   if (!visited_instrs.emplace(instr->unique_id()).second) {
-    return false;
+    return std::nullopt;
   }
+  subgraph.emplace_back(const_cast<HloInstruction *>(instr));
```

Comment:
I would move this const_cast to the two lambda functions where you call `IsSupportedF8Pattern`, and make FindF8SubgraphRecursive and IsSupportedF8Pattern take non-const parameters. This way, it's clear that when const_cast is called, the `HloInstruction*` is not actual const. E.g., do the following in the lambda functions:

```c++
m::Op().WithPredicate([&](const HloInstruction *instr) {
  return IsSupportedF8Pattern(const_cast<HloInstruction *>(instr), a, a_scale, a_mult_scale,
                              a_unary_ops);
```

const_cast is generally bad, but I don't see an elegant way of getting rid of it completely